pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, LinearCombination, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_supported_obs = frozenset(["PauliZ", "Identity"])
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_measurements(tape, supported_base_obs=("PauliZ", "Identity")):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
+
+    Args:
+        tape (QNode or QuantumScript or Callable): The quantum circuit to modify the measurements of.
+        supported_base_obs (Optional, Iterable(Str)): A list of names of supported base observables.
+            Allowed names are 'PauliX', 'PauliY', 'PauliZ' and 'Hadamard'. If no list is provided,
+            the transform will diagonalize everything.
+
+    Returns:
+        qnode (QNode) or tuple[List[QuantumScript], function]: The transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    .. note::
+        This transform will raise an error if it encounters non-commuting terms. To avoid non-commuting terms in
+        circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
+        can be applied.
+
+
+    **Examples:**
+
+    This transform allows us to transform QNode measurements into the measurement basis by adding
+    the relevant diagonalizing gates to the end of the tape operations.
+
+    .. code-block:: python3
+
+        from pennylane.transforms import diagonalize_measurements
+
+        dev = qml.device("default.qubit", wires=2)
```

Comment:
```suggestion
        dev = qml.device("default.qubit")
```
Wires is no longer needed in `default.qubit`. It's not a big deal, I just prefer the majority of examples to not use `wires` unless needed.